Storytellers for Dynamic

Putnam3145 opened the pull request:
## About The Pull Request

Replaces calm/chaos dynamic voting with storytellers: basically an extra way to weigh antags/events based on various criteria, with storytellers being things that sort using those criteria.

Also implements approval voting and ranked choice for votes, depending on arguments. Storytellers use Schulze scoring of ranked choice ballots by default. Hopefully the uses of these can be expanded over time.

## Why It's Good For The Game

It makes dynamic voting less dichotomous, allowing for more people to get what they want. Implementation of approval voting and ranked choice could also help a bit with stuff; it would be rather easy to swap map voting to approval voting or ranked choice this way, for example.

## Changelog
:cl:
add: Dynamic storytellers, a new voting paradigm for dynamic
add: Support for approval voting and condorcet (ranked choice) voting in server votes
/:cl:

Putnam3145 committed:
that's storytellers. time for voting

Putnam3145 committed:
voting makes everything dumb

Putnam3145 committed:
actually i still need this

Putnam3145 committed:
alright that's that

Putnam3145 committed:
uh, vote changes necessary too

Putnam3145 committed:
linter is correct

Ghommie commented on code/game/gamemodes/dynamic/dynamic_storytellers.dm:
```diff
@@ -0,0 +1,127 @@
+/datum/dynamic_storyteller/cowabunga
+	name = "Cowabunga"
+	curve_centre = 10
+	desc = "The most chaotic rules only. A validhunter's dream."
```
Saying "A validhunter's dream." is pretty much a sucker to those who complain about antagonists being hunted too much. And considering this also drafts in some potentially dangerous random events, I guess the point is more of a wild ride than a witch hunt.
```suggestion
	desc = "The most chaotic rules only. Expect the unexpectable."
```
P.S. The suggested description is just an example, and a reference to some ugandan meme. 

I'll perhaps review the rest later.

dzahlus replied:
Just change it to "Cowabunga it is"

Putnam3145 committed:
make cowabunga description less, uh, bad

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into storytellers

Putnam3145 committed:
Added bloodsucker weights.

Putnam3145 committed:
Made intrigue not care about valid.

Putnam3145 committed:
whoops

Putnam3145 committed:
workin on IRV

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into storytellers

Putnam3145 committed:
it works

Putnam3145 committed:
why did i not have revs as conversion

Putnam3145 committed:
thanks linter

Putnam3145 committed:
aaa

Putnam3145 committed:
actually condorcet not IRV

Putnam3145 committed:
screw it. minimax

Putnam3145 committed:
no need for this much whitespace

Putnam3145 committed:
Merge branch 'master' of https://github.com/Citadel-Station-13/Citadel-Station-13 into storytellers

Putnam3145 committed:
Added statbus support

Putnam3145 committed:
i believe they call this a "weh"

Putnam3145 committed:
Made vote not fire every tick, added "force" option that pops as soon as vote starts

Putnam3145 committed:
how one = can make a difference

Putnam3145 committed:
Added extended, deobfuscated, made "how many to vote" bigger

Putnam3145 committed:
make vote last longer, too

Putnam3145 committed:
fixes

Putnam3145 committed:
made it nag you a bit, added logging

Putnam3145 committed:
better stats

Putnam3145 committed:
better logging for why candidates aren't working

Putnam3145 committed:
fixed overrefunds (again)

Putnam3145 committed:
god i'm removing this extension completely

Putnam3145 committed:
made toggle more indicative

Putnam3145 committed:
nagging people is annoying

Putnam3145 committed:
fixed the logging

Putnam3145 committed:
clarity

Putnam3145 committed:
haha WHOOPS

Putnam3145 committed:
renamed the storytellers and added secret to voting

Ghommie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -116,17 +169,24 @@ SUBSYSTEM_DEF(vote)
+				SSblackbox.record_feedback("nested tally","voting",choices[choices[i]],list(vote_title_text,choices[i]))
+		if(RANKED_CHOICE_VOTING)
+			for(var/i=1,i<=voted.len,i++)
+				for(var/j=1,j<=voted[voted[i]].len,j++)
```
Linter check. define a list L for voted[voted[i]].

Putnam3145 committed:
WHOOP FUCK

Putnam3145 committed:
more logging

Putnam3145 committed:
linter u right

Putnam3145 committed:
no assassination at all (!!) on calm

Ghommie commented on code/__DEFINES/citadel_defines.dm:
```diff
@@ -119,3 +119,11 @@
+#define RANKED_CHOICE_VOTING 2
+
+// Dynamic stuff (see above, move later)
+#define NO_ASSASSIN (1<<0)
```
I don't like citadel_defines.dm, it's some cringe. :grimacing: 
If you could move these into their appropriate defines files , I'd appreciate it.

Putnam3145 replied:
doing it

Ghommie commented on code/controllers/subsystem/vote.dm:
```diff
@@ -196,27 +243,51 @@ SUBSYSTEM_DEF(vote)
-			voted[usr.ckey] = vote
-			choices[choices[vote]]++
-			return vote
+		if(vote && vote >= 1 && vote <= choices.len)
```
```suggestion
		if(ISINRANGE(vote, 1, choices.len))
```

Ghommie commented on code/game/gamemodes/dynamic/dynamic_storytellers.dm:
```diff
@@ -0,0 +1,149 @@
+/datum/dynamic_storyteller/cowabunga
+	name = "Chaotic"
+	curve_centre = 10
+	desc = "Chaos: high. Variation: high. Likely antags: clock cult, revs, wizard."
```
In a future, these sort of descriptions will have to be actually classified on New() based on factors such as their property weights and accessible rulesets for easier maintenance, not gonna lie.

Putnam3145 replied:
not entirely sure i want them permanently at all

Ghommie reviewed (approved):
I heard there have been a few issues with the testmerge in the past days which should have been resolved by now, so tell me when you think it's ready for merge.

Putnam3145 committed:
Con vey ance

Putnam3145 committed:
yeah conveyance

Putnam3145 committed:
reobfuscated dynamic vote

Putnam3145 committed:
Storytellers do more things now

Putnam3145 committed:
vote overhaul start

Putnam3145 committed:
Revert "vote overhaul start"

This reverts commit 232811dd18c58b8e6e52228a4ca0486d08c0ac2b.

Putnam3145 committed:
logging!

Putnam3145 committed:
better make that logging work

Putnam3145 committed:
also make weight adjustment round-to-round work, probs

Putnam3145 committed:
Added storyteller to dynamic game mode panel.

Putnam3145 committed:
Made ion storm not THE ONLY EVENT THAT WILL RUN ON STORY

Putnam3145 committed:
made it Taller

Putnam3145 committed:
and a line break goes there

Putnam3145 committed:
chaotic shouldn't have so many events

Putnam3145 committed:
Allowed preventing dynamic storytellers if too common recently

Putnam3145 committed:
while we're at it let's make it work for all of 'em

Putnam3145 committed:
WOW WHOOPS

Putnam3145 committed:
v o t e s a v i n g

Putnam3145 committed:
Renamed "Classic" to "Random".

Putnam3145 committed:
Brought the weight up on extended/chaotic

Putnam3145 committed:
Made war ops always allowed on teamwork, chaotic

Putnam3145 committed:
wow how'd i leave that in

Putnam3145 committed:
Fixes a runtime, almost definitely

Putnam3145 committed:
WOW THAT'S A BIG NUMBER

Putnam3145 committed:
balancing is good

Putnam3145 committed:
Made story *purely* story, no extended-weighting

Putnam3145 committed:
actually screw this "remove if recent" thing it's too much trouble

Putnam3145 committed:
Rebalancing bloodsuckers a tad

Putnam3145 committed:
SCORING WRONG THE WHOLE TIME

Putnam3145 committed:
made "random" more random

Putnam3145 committed:
Made no antags show up after auto call

Putnam3145 committed:
Code improvements

Putnam3145 committed:
Halved midround antag injection chance on calm

Putnam3145 committed:
caps

Putnam3145 committed:
Balance pass (see description)

1. Made bloodsuckers more likely on calm, story, slightly more common on intrigue
2. Made intrigue no longer have low threat by default
3. Made abductors have lower weights in general, making them less likely to show up
4. Made sentient disease, revenant less likely on calm
5. Added a new storyteller, "Conversion", which rolls conversion antags.

Putnam3145 committed:
Made loading saved votes not automatic

Putnam3145 committed:
majority judgement time

Putnam3145 committed:
1-5 -> -2 to 2

Putnam3145 committed:
2 minute vote time

Putnam3145 committed:
more conveyance on how the voting works

Putnam3145 committed:
try recent round removal again

Putnam3145 committed:
removed extended weighting on intrigue

Putnam3145 committed:
more extreme.

Putnam3145 committed:
Made random always available.

Putnam3145 committed:
Renamed the options to make clear this isn't scoring?

Putnam3145 committed:
also added individual ballot gathering

Putnam3145 committed:
tweaks

Putnam3145 committed:
moved a thing to prevent a runtime

Putnam3145 committed:
wow good thing this was impossible to check

Putnam3145 committed:
actually no

Putnam3145 committed:
saving is back

Putnam3145 committed:
fixed voting.

FUCK
